#+TITLE: Graduation Timeline
#+SETUPFILE: setup.org
#+OPTIONS: toc:nil p:t pri:t tex:t

 I've put down times for when I would like to have certain items finished (~DEADLINE~). I believe the timeline is aggressive but achievable.
 ~TODO~ items are highlighted in red, and where there are multiple items in a category, they are prioritized by letter, (with "A" being the highest)


* Functional Genetics by Expectation Maximization Project 
  My plan is for this chapter to consist of a submitted paper (targeting Bioinformatics).
  DEADLINE: <2020-05-27 Wed>
*** TODO Finish analyses
    DEADLINE: <2020-05-22 Fri>
**** DONE [#A] Run FGEM on putative cancer genes from driverMAPS paper 
     DEADLINE: <2020-05-07 Thu>
I ran FGEM on 20 lists of gene-level Bayes Factors obtained from cite:zhao19_detail_model_posit_selec_improv
**** TODO [#A] Validate FGEM features against COSMIC cancer gene census 
DEADLINE: <2020-05-21 Thu>
I will validate the set of gene-level features I identify by estimating enrichment in the COSMIC cancer gene census
**** DONE [#B] Make a decision about feature collinearity issue
     DEADLINE: <2020-05-18 Mon>
If the features from genes are highly correlated, the parameter estimates can be unstable. The current plan is to 
detect and merge or remove highly correlated features from the analysis
*** TODO Finish first draft
    DEADLINE: <2020-05-18 Mon>
I believe I can have a first draft finished by this date.  Obviously work on the draft will happen concurrently with analysis work.
* Polygenic RSS
As it stands there will probably more than one paper's worth of work to be written up for this project
by the time it is finished. The current plan is for this project is to get one paper out of the project
and for the rest of the work to go in the dissertation.
** Write Paper 
   DEADLINE: <2020-06-20 Sat>
*** TODO Finish analyses
    DEADLINE: <2020-06-04 Thu>
**** DONE [#A] Simulate UK Biobank GWAS with polygenic genetic architecture
**** TODO [#A] Simulate UK Biobank GWAS with sparse genetic architecture
**** TODO [#B] Finish UK-biobank/Genbank GWAS real-data analysis

# ***** TODO [#A]  Complete with vs without ldetect/ldshrink single-chromosome simulations 

# ***** TODO Need to decide if I want to use in-sample LD, or 1000 genomes out-of-sample LD (for apples to apples comparison with LDSC)
# Several UKB LDSC results with 1KG LD are already available so it might not be worth the effort of rerunning
# **** TODO [#B] Figure out what to do about inflated $\hat{h^2}$ estimate at high $h^2$
     
#      I have a few hypotheses about what's causing the inflation:

# 1) Underestimating local LD due to ldshrink
# 2) Underestimating long-range LD due to ldetect blocking
# 3) Over-estimating intermediate-range LD due to ldetect blocking
# 4) Breakdown of assumptions of RSS likelihood

* Preterm Birth Fine Mapping
  DEADLINE: <2020-06-30 Tue>
My plan is to write up the analyses that I've already done (that went in to the first March of Dimes paper) as my third chapter.  Depending on how things go, additional 
analyses that need to be done can go in to another paper or simply into the chapter.
** TODO [#A] Write up analyses for chapter
   DEADLINE: <2020-06-28 Sun>   
** TODO [#B] Finish additional analyses
*** TODO [#C] Incorporate TCM term vs preterm ATAC annotations into a new model
*** TODO [#D] Incorporate TF ChIP-seq annotations into a new model
*** TODO [#D] Compare which loci are implicated by which models

bibliographystyle:unsrt
bibliography:references.bib


