<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-04-30 Thu 09:41 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Functional Enrichment by Expectation Maximization</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Nicholas Knoblauch">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
<link rel="stylesheet" type="text/css" href="https://gongzhitaao.org/orgcss/org.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2019 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Functional Enrichment by Expectation Maximization</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org31e11cb">Overview</a></li>
<li><a href="#org07da710">Model</a></li>
</ul>
</div>
</div>

<div id="outline-container-org31e11cb" class="outline-2">
<h2 id="org31e11cb">Overview</h2>
<div class="outline-text-2" id="text-org31e11cb">
<p>
Combine gene-level Bayes-Factors from whole-exome sequencing studies with pre-existing biological knowledge in the form of genic and genomic annotation to identify causal genes and enriched pathways. 
</p>
</div>
<div id="outline-container-orgad52fa0" class="outline-4">
<h4 id="orgad52fa0">Rationale</h4>
<div class="outline-text-4" id="text-orgad52fa0">
<p>
Preliminary results using data from an Autism Spectrum Disorder implicate previously identified and novel genes and annotations as being disease-related.  In ASD, transmitted and <i>de novo</i> variants are known to play a role in disease risk.  Among the top associated annotations are  several Gene Ontology terms related to neural development, and features related to mutational constraint (as defined by the ExAC missense Z score). 
The combination of gene-level summaries of phenotypic association with relevant gene-level annotations can dramatically improve in the discrimination betweeen causal genes and pathways from non-causal ones.  We further hypothesize that explicitly modeling the inherently hierarchical nature of many gene-level annotations can further improve classification.  We believe that in the case of disease, that features related to mutational constraint,relevant tissue-level expression, and phylogenetic conservation will be enriched functional classes.
</p>
</div>
</div>
<div id="outline-container-orgbb0f9a3" class="outline-4">
<h4 id="orgbb0f9a3">Genomic Annotation</h4>
<div class="outline-text-4" id="text-orgbb0f9a3">
<p>
The gene-level annotations will come from several sources.  The Exome Aggregation Consortiom (ExAC) has created a list of gene-level constraint Z scores for every gene in the genome.  The constraint scores are based on the number of observed non-synonymous mutational events per-gene versus what would be expected from the length of the gene and the local mutation rate. The Gene Ontology Consortium maintains a database describing genes in a hierarchical manner according to their molecular function as well as  which biological processes they are thought to be involved in. Genes associated with a common biological process can be considered members of a gene set, which is a kind of binary gene-level annotation.  There also exist a large number of pre-existing datasets that can be used to create gene-level annotation.  The results of gene-expression profiling experiments can be summarised  by annotating a gene according to some summary of its expression level relative to other genes, or if differential gene-expression has been investigated in a context deemed relevant, a summary of that differential expression in the form of a gene-level Z-score can be incorporated.  More generally, the results of previous experiments can be summarised as a list of candidate genes.  
</p>
</div>
</div>
<div id="outline-container-orgedb8d96" class="outline-4">
<h4 id="orgedb8d96">Model Outline</h4>
<div class="outline-text-4" id="text-orgedb8d96">
<p>
For each gene, let \(Z_i=1\) indicate that this gene is "causal" for our phenotype of interest, and \(Z_i=0\) if it is not. If we have a gene-level model of whether or not a gene is a causal gene, we can define a gene-level Bayes factor \(B_i=\frac{P(x_i|Z_i=1)}{P(x_i|Z_i=0)}\).   The prior probability that a gene is causal \(P(Z_i=1)=\pi_i\) depends on a vector of gene-level annotations, \(A_i\), and a vector of effects of each feature \(\beta\):
</p>

<p>
\[\text{logit}(P(Z_i=1))=A_i \beta_i\] 
</p>

<p>
Conditioning on \(\beta\),or alternatively conditioning on \(pi_i\): (remember \(\pi\) is a function of \(\beta\), which is not known, and \(A\), which is known) 
\[P(x|\pi) \propto \prod_i [\pi_i B_i+(1-\pi)]\]
We can use the Expectation Maximization algorithm to estimate the posterior probability that \(Z_i=1\), as well as the $&beta;$s for each annotation
</p>

<p>
The joint probability of both \(Z_i\) and \(x_i\) conditioning on \(\beta\) is :
\[P(x_i,Z_i|\pi_i)=P(Z_i|\pi_i)P(x_i|Z_i)=\pi_i^{Z_i}(1-\pi)^{1-Z_i}B_i^{Z_i}\]
The <b>E</b> Step is the expected value of the log likelihood function:
\[Q(\beta|\beta^{(t)})=\sum_i \sum_{Z_i} P(Z_i|x_i,\beta^{(t)} \log(P(x,Z_i|\beta))\]
This is the expectation of \(Z\) for each gene.  We can then find the \(\beta\) that maximizes the likelihood at the current estimate of \(Z\).
</p>

<p>
If we start with some reasonable guess for \(Z\) (like having \(P(Z_i=1)\) be some function of \(B_i\)), we can calculate the posterior probability and estimates of \(\beta\).   
</p>
</div>
</div>
</div>


<div id="outline-container-org07da710" class="outline-2">
<h2 id="org07da710">Model</h2>
<div class="outline-text-2" id="text-org07da710">
<p>
Suppose we have gathered genetic data for a set of individuals to identify which genes are causally related to a disease or trait of interest.  For each gene \(g \in \{1 \dots G\}\), let the indicator variable \(Z_g=1\) indicate that
gene \(g\) is causaully related to our trait of interest.  We can summarise the evidence for and against our hypothesis that \(Z_g=1\) using a bayes factor:
</p>

<p>
\[B_g=\frac{P(x_g|Z_g=1)}{P(x_g|Z_g=0)}\]
</p>

<p>
where \(x_g\) is the aforementioned genetic data for gene \(g\).
</p>

<p>
Suppose we know a set of \(F\) functional annotations or properties for each of our \(G\) genes, such that \(A_{f,g}\) with  is the value of annotation \(f\) for gene \(g\). Let us define the vector \(\boldsymbol{\beta}\) and the function 
\(pi(\boldsymbol{\beta},\textbf{A})\) such that satisfies the following conditional probability that \(Z_g=1\)
</p>

<p>
\[P(Z_g=1|\textbf{A}_{.,g},\boldsymbol{\beta}) = \frac{1}{1+e^{-(\beta_{0}+\sum_{f=1}^F{A_{f,g}\beta_f})}} = \pi(\boldsymbol{\beta},\textbf{A}_{.,g})\]
</p>


<p>
\[ P(\textbf{x}|\boldsymbol{\beta},\textbf{A})=\prod_{g=1}^{G}P(x_g|\boldsymbol{\beta})=\prod_{g=1}^{G}[\pi(\boldsymbol{\beta},\textbf{A}_{.,g}) P(x_g|Z_g=1)+(1-\pi(\boldsymbol{\beta},\textbf{A}_{.,g}))P(x_g|Z_g=0)]\]
</p>

<p>
By factorizing out the term \(\prod_{g=1}^{G} P(x_g|Z_g=0)\) which does not depend on \(\boldsymbol{\beta}\), we can express the likelihood for \(\boldsymbol{\beta}\) in terms of \(\textbf{B}\):
</p>


<p>
\[P(\textbf{x}|\boldsymbol{\beta},\textbf{A}) \propto \prod_{g=1}^{G}[\pi(\boldsymbol{\beta},\textbf{A}_{.,g})B_g+(1-\pi(\boldsymbol{\beta},\textbf{A}_{.,g}))]\]
</p>


<p>
Our goal is both to estimate \(\boldsymbol{\beta}\) as well as \(P(Z_g=1|\textbf{A}_{.,g},\boldsymbol{\beta},x_g)\) for each gene.  
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Nicholas Knoblauch</p>
<p class="date">Created: 2020-04-30 Thu 09:41</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
