<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-07-30 Thu 11:26 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Functional Enrichment by Expectation Maximization</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Nicholas Knoblauch">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
<link rel="stylesheet" type="text/css" href="https://gongzhitaao.org/orgcss/org.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2019 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Functional Enrichment by Expectation Maximization</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org7c05cae">Introduction</a></li>
<li><a href="#org7dc987a">Method</a>
<ul>
<li><a href="#org07da710">Model</a></li>
<li><a href="#org5facbbf">Validation against intOGen</a></li>
<li><a href="#orgac60054">Data</a></li>
<li><a href="#org3d7e31d">Software</a></li>
</ul>
</li>
<li><a href="#org37638e0">Results</a>
<ul>
<li><a href="#org3fca551">FGEM identifies biologically relevant gene-level annotations</a></li>
<li><a href="#org3d9b10f">FGEM integrates multiple gene-level annotations to reprioritize genes</a></li>
<li><a href="#orgd2b0148"></a></li>
<li><a href="#orgf83eeb8">Highly enriched single features</a></li>
</ul>
</li>
<li><a href="#org5c9eeca">Discussion</a>
<ul>
<li><a href="#org90142d2">Consequences of the latent variable approach</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org7c05cae" class="outline-2">
<h2 id="org7c05cae">Introduction</h2>
<div class="outline-text-2" id="text-org7c05cae">
<p>
When interpreting the results of a genome-scale genetic association study, be it a GWAS, a whole exome sequencing study, or a whole genome sequencing study, there are two questions that are of interest:
</p>
<ol class="org-ol">
<li>which genes are associated with the trait of interest, and 2) what are the common properties of the genes that contribute to disease risk. The answer to one of these questions can inform the other.</li>
</ol>
<p>
Obviously, identifying a set of properties common to causal genes is made much easier if the set of causal genes is known with certainty. Furthermore, if the set of causal genes is not known, but
prior knowledge about the properties of the causal genes is available, that prior knowledge can be incorporated to implicate genes that might otherwise not have been identified due to a lack of statistical power.
Prior information can also be used to deprioritize genes that might otherwise be implicated, despite being false positives.  Despite the complementarity of these 
</p>
</div>
</div>



<div id="outline-container-org7dc987a" class="outline-2">
<h2 id="org7dc987a">Method</h2>
<div class="outline-text-2" id="text-org7dc987a">
</div>
<div id="outline-container-org07da710" class="outline-3">
<h3 id="org07da710">Model</h3>
<div class="outline-text-3" id="text-org07da710">
<p>
Suppose we have gathered genetic data for a set of individuals to identify which genes are causally related to a disease or trait of interest. For each gene \(g \in \{1 \dots G\}\), let the indicator variable \(z_g=1\) indicate that
gene \(g\) is causaully related to our trait or disease of interest.  We can summarise the evidence for and against our hypothesis that \(z_g=1\) using a bayes factor:
</p>

<p>
\[B_g=\frac{P(x_g|z_g=1)}{P(x_g|z_g=0)}\]
</p>

<p>
where \(x_g\) is the subset of the aforementioned genetic data corresponding to the $g$th gene.
</p>

<p>
Suppose further that we know a set of \(F\) functional annotations or properties for each of our \(G\) genes.  Let \(\textbf{a}_g\) denote the length \(F\) vector of annotations for gene \(g\), and \(\textbf{A}\) denote the matrix with \(F\) rows and \(G\) 
columns consisting of \(\textbf{a}_1 ...  \textbf{a}_G\)
We define the vector \(\boldsymbol{\beta}\) and the function \(\pi(\boldsymbol{\beta},\textbf{a}_g)\) such that:
</p>

<p>
\[\pi(\boldsymbol{\beta},\textbf{a}_g) =  \frac{1}{1+e^{-(\beta_{0}+\sum_{f=1}^F{A_{f,g}\beta_f})}} =  P(z_g=1|\textbf{a}_g,\boldsymbol{\beta})\]
</p>

<p>
We can compute the likelihood of a particular value of \(\boldsymbol{\beta}\) by treating the data from each gene as coming from a two component mixture model (where \(z_g=1\) and where \(z_g=0\)) and marginalizing over the two components
</p>

<p>
\[ P(\textbf{x}|\boldsymbol{\beta},\textbf{A})=\prod_{g=1}^{G}P(x_g|\boldsymbol{\beta})=\prod_{g=1}^{G}[\pi(\boldsymbol{\beta},\textbf{a}_g) P(x_g|z_g=1)+(1-\pi(\boldsymbol{\beta},\textbf{a}_g))P(x_g|z_g=0)]\]
</p>

<p>
By factorizing out the term \(\prod_{g=1}^{G} P(x_g|z_g=0)\) which does not depend on \(\boldsymbol{\beta}\), we can express the likelihood for \(\boldsymbol{\beta}\) in terms of \(\textbf{B}\):
</p>


<p>
\[P(\textbf{x}|\boldsymbol{\beta},\textbf{A}) \propto \prod_{g=1}^{G}[\pi(\boldsymbol{\beta},\textbf{a}_g)B_g+(1-\pi(\boldsymbol{\beta},\textbf{a}_g))]\]
</p>

<p>
Given a particular value of \(\boldsymbol{\beta}\), and a bayes factor \(B_g\), we arrive at a new posterior probability that \(z_g=1\):
</p>

<p>
\[P(z_g=1 | B_g, \boldsymbol{\beta},\textbf{a}_g) = \frac{\pi(\boldsymbol{\beta},\textbf{a}_g) B_g}{\pi(\boldsymbol{\beta} , \textbf{a}_g) B_g + 1 - \pi(\boldsymbol{\beta},\textbf{a}_g)}\]
</p>


<p>
Our goal is both to estimate \(\boldsymbol{\beta}\) for a relevant set of features as well as \(P(Z_g=1|\textbf{a}_g,\boldsymbol{\beta},x_g)\) for each gene.  We use a penalized maximum-likelihood
 approach to estimate \(\boldsymbol{\beta}\), from which we then \(P(Z_g=1|\textbf{a}_g,\boldsymbol{\beta},x_g)\). The number of gene-level features one might include in such a model is very large.  It is impossible, from both a computability 
and interpretability standpoint, to include all conceivable features in the model. A related but distinct issue is that of collinearity.
As the number of features in the model increases, the probability that some subset of features will be collinear with one-another increases,
which can complicate model-fitting, as \(\beta\) becomes unidentifiable. This is especially important when a binary, hierarchical feature set like the Gene Ontology.To avoid this issue, 
we employ a multi-stage model fitting procedure. 
In the first step, all single-feature-plus-intercept models are fit, and a $p$-value is obtained for each model byt comparing to the intercept-only model via the likelihood ratio test.  
</p>

<p>
From this set of single-feature models, all of the nonsignificant (i.e features with Benjamini-Hochberg adjusted $p$-values greater than 0.05) univariate features for each cancer type
were removed from the analysis.  
</p>
</div>
<div id="outline-container-org11c3e99" class="outline-4">
<h4 id="org11c3e99">Joint Model</h4>
<div class="outline-text-4" id="text-org11c3e99">
<p>
Significant features for each cancer type were then combined in a joint model and fit by maximizing the marginalized likelihood, penalized with an elastic-net penalty.  
</p>

<p>
The Limited Memory Broyden-Fletcher-Goldfarb-Shanno algorithm (LM-BFGS) (<a href="#citeproc_bib_item_2">Byrd et al. 1995</a>)  is among the most popular algorithms 
for unconstrained optimization over scalar, differentiable functions. 
</p>

<p>
One limitation of LM-BFGS, is that the function that is being optimized must be differentiable.  Unfortunately, sparsity-inducing $l<sub>1</sub>$-regularized models of the form
</p>

<p>
\[f(\boldsymbol{\theta})=p(\boldsymbol{\theta} | \textbf{x}) + C \Vert \boldsymbol{\theta} \Vert_1\] are not differentiable when any of the elements of the parameter vector \(\boldsybmol{\theta}\) are 0.  The Orthant-wise limited-memory quasi-Newton method is a variant of LM-BFGS which is designed precisely for 
fitting $l<sub>1</sub>$-regularized, sparsity inducing models. 
</p>
</div>
</div>

<div id="outline-container-orgd6f402b" class="outline-4">
<h4 id="orgd6f402b">Regularization by (relaxed) elastic-net</h4>
<div class="outline-text-4" id="text-orgd6f402b">
<p>
Our multivariate fitting procedure consists of two steps.  The first emphasizes feature selection, and the second emphasizes shrinkage.  In the first step, we minimize the objective function corresponding to the negative of the elastic-net penalized log-likelihood:
\[ -\mathcal{L}(\textbf{x}|\boldsymbol{\beta},\textbf{A}) + \lambda\left( \frac{1-\alpha}{2} \sum_{j=2}^F\beta_j^2 + \alpha \sum_{j=2}^F|\beta_j| \right) \]
</p>

<p>
Where \(\mathcal{L}(\textbf{x}|\boldsymbol{\beta},\textbf{A}) = \sum_{g=1}^{G}[\log\left(\pi(\boldsymbol{\beta},\textbf{a}_g)B_g+(1-\pi(\boldsymbol{\beta},\textbf{a}_g))\right)]\).
</p>

<p>
The overall level of sparsity in the model is controlled by the parameter \(\lambda\), while the proportion of \(l_1\) vs \(l_2\) penalty is determined by \(\alpha\).  As a default, the value of \(\alpha\) is fixed at 0.85, and \(\lambda\) 
takes on 100 values starting at zero, and ending at \(\lambda_{\text{max}}\), on a log scale.  \(\lambda_{\text{max}}\) is defined as the smallest value of \(\lambda\) for which the objective function is minimized, and all coefficients 
but the intercept are 0 (the intercept term is not subject to the elastic net penalty). The optimal value of \(\lambda\) is chosen using 10-fold cross-validation.  
</p>

<p>
In the second step, features with \(\beta_f=0\) are removed from the analysis, and cross-validation over 100 values of \(\lambda\) occurs again, this time with \(\alpha = 0\), meaning there is no \(l1\) penalty.  The motivation for this two-stage
procedure comes from the "relaxed" lasso (<a href="#citeproc_bib_item_6">Hastie, Tibshirani, and Tibshirani 2017</a>), wherein the lasso is used for feature selection, and the model is refit without a lasso penalty.
</p>
</div>
</div>

<div id="outline-container-org1175ff8" class="outline-4">
<h4 id="org1175ff8">Removal of highly enriched features</h4>
<div class="outline-text-4" id="text-org1175ff8">
<p>
    A subset of features were remvoed after the univariate feature-selection stage as they exhibited enrichment estimates in excess of credulity. Features which corresponded to posterior probabilities in excess of 0.3 were 
recorded and excluded from consideration for inclusion in the joint model (they were however included when calculating the FDR-adjusted p-values for the univariate tests).  
</p>
</div>
</div>

<div id="outline-container-org5613966" class="outline-4">
<h4 id="org5613966">Comparison with Fisher's Exact test</h4>
<div class="outline-text-4" id="text-org5613966">
<p>
In the case of a single binary feature, one can apply a Bayes Factor cutoff to obtain a contingency table and assess the enrichment of the feature using Fisher's Exact test.
 We compared FGEM with Fisher's exact test, using an FDR cutoff of 0.1, and compared the $p$-values to those obtained from the single-feature, FGEM the likelihood ratio test $p$-values.
</p>
</div>
</div>
</div>

<div id="outline-container-org5facbbf" class="outline-3">
<h3 id="org5facbbf">Validation against intOGen</h3>
<div class="outline-text-3" id="text-org5facbbf">
<p>
IntOGen is a database of cancer driver genes (<a href="#citeproc_bib_item_5">Gonzalez-Perez et al. 2013</a>).  It is populated by an ensemble method that incorporates seven different methods for identifying cancer driver genes.  
It weights each of the 7 methods according to their ability to predict membership in the The Catalogue of Somatic Mutations in Cancer (COSMIC) Cancer Gene Census (CGC) (<a href="#citeproc_bib_item_7">Sondka et al. 2018</a>).
To validate the FGEM models, we compared the posterior under the functional model to the functional posterior under the uniform model.  In every cancer type, validated cancer genes had a higher functional posterior than 
uniform posterior, while genes that were not previously identified as cancer genes had on average lower functional posterior compared to uniform (see table). 
</p>
</div>
</div>

<div id="outline-container-orgac60054" class="outline-3">
<h3 id="orgac60054">Data</h3>
<div class="outline-text-3" id="text-orgac60054">
</div>

<div id="outline-container-org535aaa0" class="outline-4">
<h4 id="org535aaa0">Gene-Level Summary Statistics</h4>
<div class="outline-text-4" id="text-org535aaa0">
<p>
A set of Bayes factors from a study of cancer driver genes (<a href="#citeproc_bib_item_10">Zhao et al. 2019</a>).  For each of 20 TCGA tumor types, roughly 20,000 genes were analyzed and the 
posterior probability that each gene (in each cancer type) was a causal gene was assessed and summarized via Bayes Factor.  
</p>
</div>
</div>

<div id="outline-container-org353cc6e" class="outline-4">
<h4 id="org353cc6e">Gene-Level Annotations</h4>
<div class="outline-text-4" id="text-org353cc6e">
</div>
<ul class="org-ul">
<li><a id="orgf07b23c"></a>Gene Ontology<br>
<div class="outline-text-5" id="text-orgf07b23c">
<p>
The "Biological Process" Gene Ontology (<a href="#citeproc_bib_item_9">The Gene Ontology Consortium 2018</a>)  was downloaded from the Bioconductor package <code>GO.db</code> (<a href="#citeproc_bib_item_3">Carlson 2020</a>). Of the 10,930 possible biological process gene ontology terms, the 2,198 terms that 
include 10 or more genes were analyzed, so as to reduce the multiple testing burden.  
</p>
</div>
</li>
</ul>
</div>
</div>


<div id="outline-container-org3d7e31d" class="outline-3">
<h3 id="org3d7e31d">Software</h3>
<div class="outline-text-3" id="text-org3d7e31d">
<p>
Our method is distributed as a freely available R package <a href="https://github.com/CreRecombinase/FGEM">FGEM</a> (<a href="#citeproc_bib_item_8">Team 2020</a>) .FGEM relies on the <code>SQUAREM</code> package to accelerate EM convergence  (<a href="#citeproc_bib_item_4">Du and Varadhan 2020</a>)   
and the hot-path functions are implemented as C++ functions using the <code>RcppEigen</code> (<a href="#citeproc_bib_item_1">Bates and Eddelbuettel 2013</a>) package.  
</p>
</div>
</div>
</div>


<div id="outline-container-org37638e0" class="outline-2">
<h2 id="org37638e0">Results</h2>
<div class="outline-text-2" id="text-org37638e0">

<div class="figure">
<p><img src="./fisher_vs_fgem.png" alt="fisher_vs_fgem.png">
</p>
</div>
</div>

<div id="outline-container-org3fca551" class="outline-3">
<h3 id="org3fca551">FGEM identifies biologically relevant gene-level annotations</h3>
</div>

<div id="outline-container-org3d9b10f" class="outline-3">
<h3 id="org3d9b10f">FGEM integrates multiple gene-level annotations to reprioritize genes</h3>
<div class="outline-text-3" id="text-org3d9b10f">

<div class="figure">
<p><img src="./fgem_posterior_plot.png" alt="fgem_posterior_plot.png">
</p>
</div>


<div class="figure">
<p><img src="./fgem_enrichment_plot.png" alt="fgem_enrichment_plot.png">
</p>
</div>
</div>
</div>



<div id="outline-container-orgd2b0148" class="outline-3">
<h3 id="orgd2b0148"></h3>
</div>

<div id="outline-container-orgf83eeb8" class="outline-3">
<h3 id="orgf83eeb8">Highly enriched single features</h3>
<div class="outline-text-3" id="text-orgf83eeb8">
<p>
As stated in the Methods section, Features which corresponded to posterior probabilities in excess of 0.3 were excluded from consideration for inclusion in the joint model.  One consequence of FGEM's latent variable approach
is that it is possible to encounter a scenario analogous to the "separation" problem in logistic regression.  In logistic regression, if one (or several) predictors predict an output perfectly, the likelihood for that 
feature is not maximized at a finite parameter value. Because there is high overlap in the Gene Ontology feature-set, little information is lost by excluding these problematic features.
</p>
</div>
</div>
</div>


<div id="outline-container-org5c9eeca" class="outline-2">
<h2 id="org5c9eeca">Discussion</h2>
<div class="outline-text-2" id="text-org5c9eeca">
</div>
<div id="outline-container-org90142d2" class="outline-3">
<h3 id="org90142d2">Consequences of the latent variable approach</h3>
<div class="outline-text-3" id="text-org90142d2">
<p>
The latent variable \(z_i\) in the FGEM model is by definition unobserved, 
One of the most salient features of FGEM as compared to Fisher's Exact test is that FGEM does not binarize data into significant vs insignificant.  For a particular gene set in a particular dataset,
the enrichment of Fisher's exact test is determined by the cardinality of the entries in the 2 by 2 contingency table.  In the case of Bayes factors, genes for which the evidence is <i>in favor</i> of the null
 hypothesis is treated identically to genes for which the evidence is <i>in favor</i> of the alternative hypothesis, but slightly below the significance cutoff.  
</p>




<p>

</p>

<style>.csl-entry{text-indent: -1.5em; margin-left: 1.5em;}</style><h2 class='citeproc-org-bib-h2'>Bibliography</h2>
<div class="csl-bib-body">
  <div class="csl-entry"><a name="citeproc_bib_item_1"></a>Bates, Douglas, and Dirk Eddelbuettel. 2013. “Fast and Elegant Numerical Linear Algebra Using the RcppEigen Package.” <i>Journal of Statistical Software</i> 52 (5):1–24. <a href="http://www.jstatsoft.org/v52/i05/">http://www.jstatsoft.org/v52/i05/</a>.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_2"></a>Byrd, Richard H., Peihuang Lu, Jorge Nocedal, and Ciyou Zhu. 1995. “A Limited Memory Algorithm for Bound Constrained Optimization.” <i>SIAM Journal on Scientific Computing</i> 16 (5). Society for Industrial &#38; Applied Mathematics (SIAM):1190–1208. <a href="http://dx.doi.org/10.1137/0916069">http://dx.doi.org/10.1137/0916069</a>.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_3"></a>Carlson, Marc. 2020. <i>GO.Db: A Set of Annotation Maps Describing the Entire Gene Ontology</i>.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_4"></a>Du, Yu, and Ravi Varadhan. 2020. “SQUAREM: An R Package for Off-the-Shelf Acceleration of EM, MM and Other EM-Like Monotone Algorithms.” <i>Journal of Statistical Software</i> 92 (7):1–41.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_5"></a>Gonzalez-Perez, Abel, Christian Perez-Llamas, Jordi Deu-Pons, David Tamborero, Michael P Schroeder, Alba Jene-Sanz, Alberto Santos, and Nuria Lopez-Bigas. 2013. “Intogen-Mutations Identifies Cancer Drivers Across Tumor Types.” <i>Nature Methods</i> 10 (11):1081–82. <a href="https://doi.org/10.1038/nmeth.2642">https://doi.org/10.1038/nmeth.2642</a>.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_6"></a>Hastie, Trevor, Robert Tibshirani, and Ryan J. Tibshirani. 2017. “Extended Comparisons of Best Subset Selection, Forward Stepwise Selection, and the Lasso.” <i>CoRR</i>. <a href="http://arxiv.org/abs/1707.08692v2">http://arxiv.org/abs/1707.08692v2</a>.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_7"></a>Sondka, Zbyslaw, Sally Bamford, Charlotte G. Cole, Sari A. Ward, Ian Dunham, and Simon A. Forbes. 2018. “The COSMIC Cancer Gene Census: Describing Genetic Dysfunction Across All Human Cancers.” <i>Nature Reviews Cancer</i> 18 (11). Springer Science and Business Media LLC:696–705. <a href="http://dx.doi.org/10.1038/s41568-018-0060-1">http://dx.doi.org/10.1038/s41568-018-0060-1</a>.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_8"></a>Team, R Core. 2020. <i>R: A Language and Environment for Statistical Computing</i>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_9"></a>The Gene Ontology Consortium. 2018. “The Gene Ontology Resource: 20 Years and Still GOing Strong.” <i>Nucleic Acids Research</i> 47 (D1):D330–38. <a href="https://doi.org/10.1093/nar/gky1055">https://doi.org/10.1093/nar/gky1055</a>.</div>
  <div class="csl-entry"><a name="citeproc_bib_item_10"></a>Zhao, Siming, Jun Liu, Pranav Nanga, Yuwen Liu, A. Ercument Cicek, Nicholas Knoblauch, Chuan He, Matthew Stephens, and Xin He. 2019. “Detailed Modeling of Positive Selection Improves Detection of Cancer Driver Genes.” <i>Nature Communications</i> 10 (1):3399. <a href="https://doi.org/10.1038/s41467-019-11284-9">https://doi.org/10.1038/s41467-019-11284-9</a>.</div>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Nicholas Knoblauch</p>
<p class="date">Created: 2020-07-30 Thu 11:26</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
